File: backend/app/schemas/access.py
"""Public shapes for the access-resolver service.

The frontend renders these on the tenant / site / device detail pages
and on the per-user access report. Kept separate from user.py to keep
each schema module focused on one concern.
"""

from __future__ import annotations

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, EmailStr

ScopeKind = Literal["organization", "tenant", "site", "device"]


class AccessEntryPublic(BaseModel):
    user_id: UUID
    email: EmailStr
    display_name: str | None
    is_admin: bool
    role_id: UUID | None
    role_name: str | None
    source_scope_type: ScopeKind | None
    source_scope_id: UUID | None
    source_scope_label: str | None


class AccessReportPublic(BaseModel):
    scope_type: ScopeKind
    scope_id: UUID | None
    scope_label: str
    entries: list[AccessEntryPublic]


# ---------------- Per-user map ----------------


class UserScopeGrantPublic(BaseModel):
    role_id: UUID | None
    role_name: str | None
    via_scope_type: ScopeKind | None
    via_scope_id: UUID | None
    via_scope_label: str | None


class UserAccessDevicePublic(BaseModel):
    device_id: UUID
    device_name: str
    grants: list[UserScopeGrantPublic]


class UserAccessSitePublic(BaseModel):
    site_id: UUID
    site_name: str
    grants: list[UserScopeGrantPublic]
    devices: list[UserAccessDevicePublic]


class UserAccessTenantPublic(BaseModel):
    tenant_id: UUID
    tenant_name: str
    grants: list[UserScopeGrantPublic]
    sites: list[UserAccessSitePublic]


class PermissionTuple(BaseModel):
    section: str
    action: str


class UserAccessMapPublic(BaseModel):
    user_id: UUID
    tenants: list[UserAccessTenantPublic]
    permissions: list[PermissionTuple]

File: frontend/src/lib/access.ts
import { api } from "@/lib/api";

export type ScopeKind = "organization" | "tenant" | "site" | "device";

export interface AccessEntry {
  user_id: string;
  email: string;
  display_name: string | null;
  is_admin: boolean;
  role_id: string | null;
  role_name: string | null;
  source_scope_type: ScopeKind | null;
  source_scope_id: string | null;
  source_scope_label: string | null;
}

export interface AccessReport {
  scope_type: ScopeKind;
  scope_id: string | null;
  scope_label: string;
  entries: AccessEntry[];
}

export interface UserScopeGrant {
  role_id: string | null;
  role_name: string | null;
  via_scope_type: ScopeKind | null;
  via_scope_id: string | null;
  via_scope_label: string | null;
}

export interface UserAccessDeviceNode {
  device_id: string;
  device_name: string;
  grants: UserScopeGrant[];
}

export interface UserAccessSiteNode {
  site_id: string;
  site_name: string;
  grants: UserScopeGrant[];
  devices: UserAccessDeviceNode[];
}

export interface UserAccessTenantNode {
  tenant_id: string;
  tenant_name: string;
  grants: UserScopeGrant[];
  sites: UserAccessSiteNode[];
}

export interface PermissionTuple {
  section: string;
  action: string;
}

export interface UserAccessMap {
  user_id: string;
  tenants: UserAccessTenantNode[];
  permissions: PermissionTuple[];
}

export async function fetchTenantAccess(tenantId: string): Promise<AccessReport> {
  return api.get(`access/tenant/${tenantId}`).json<AccessReport>();
}

export async function fetchSiteAccess(siteId: string): Promise<AccessReport> {
  return api.get(`access/site/${siteId}`).json<AccessReport>();
}

export async function fetchDeviceAccess(deviceId: string): Promise<AccessReport> {
  return api.get(`access/device/${deviceId}`).json<AccessReport>();
}

export async function fetchUserAccessMap(userId: string): Promise<UserAccessMap> {
  return api.get(`access/user/${userId}`).json<UserAccessMap>();
}
